### PR TITLE
Add API to build Requests

### PR DESCRIPTION
**Pull request by k-owens**

### CODE REVIEW

File: Server.Core/RequestBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+                SetRequestToNoValues();
+            }
+            else
+            {
+                ParseMessageBytes(requestMessage);
+            }
+            return request;
+        }
+
+        public void SetRequestHttpVersion(string httpVersion)
```
Comment (aaronlahey) [inline Server.Core/RequestBuilder.cs]: Is having `Request` in the name of all of these methods necessary. It _is_ the `RequestBuilder` after all.

File: Server.Core/RequestBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        {
+            request.Uri = uri;
+        }
+
+        public void SetRequestMethod(string method)
+        {
+            request.Method = method;
+        }
+
+        public void SetRequestHeaders(string[] headers)
```
Comment (aaronlahey) [inline Server.Core/RequestBuilder.cs]: Is this specific enough? Would it be more valuable to have a `AddHeader` method which takes a key and value?

File: Server.Core/RequestBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        {
+            request.Body = body;
+        }
+
+        public void SetRequestBody(string bodyMessage)
+        {
+            request.Body = Encoding.UTF8.GetBytes(bodyMessage);
+        }
+
+        private byte[] GetBodyOfMessage(int startIndex, byte[] message)
```
Comment (aaronlahey) [inline Server.Core/RequestBuilder.cs]: Why is `startIndex` required? Is that needed for _every_ use case?

File: Server.Core/RequestBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        {
+            var body = new byte[message.Length - startIndex];
+            for (var index = startIndex; index < message.Length; index++)
+            {
+                body[index-startIndex] = message[index];
+            }
+            return body;
+        }
+
+        private string[] DivideHeaders(string[] allLines)
```
Comment (aaronlahey) [inline Server.Core/RequestBuilder.cs]: This class...
1. Seems to have a couple responsibilities:
   - Providing convenience methods for setting properties on the `Request` (The traditional use of the builder pattern)
   - Parsing raw data into a request.
2. The parsing raw data aspect of this class needn't be public to the world. That's an internal detail of your server.

File: Server.Core/Response.cs [hunk 5]
```diff
@@ -6,7 +6,13 @@ public class Response
     {
         public int StatusCode { get; set; }
         public string ContentType { get; set; }
-        public MemoryStream Body { get; set; }
+        public Stream Body { get; set; }
+
+        public void SetBody(byte[] bodyMessage)
```
Comment (aaronlahey) [inline Server.Core/Response.cs]: Why have this method on `Response` but a similar method on `RequestBuilder`? There seems to be some missing symmetry.

File: Server.Core/Server.cs [hunk 6]
```diff
@@ -60,7 +60,9 @@ public void HandleClients()
         private void RespondToClient()
         {
             byte[] clientMessage = Read();
-            Request request= new Request(clientMessage);
+            Console.WriteLine(Encoding.UTF8.GetString(clientMessage));
```
Comment (aaronlahey) [inline Server.Core/Server.cs]: Is this leftover debugging?

File: Server.Core/Server.cs [hunk 7]
```diff
@@ -60,7 +60,9 @@ public void HandleClients()
...
+            RequestBuilder requestbuilder = new RequestBuilder();
+            Request request = requestbuilder.BuildRequestFromWholeMessage(clientMessage);
```
Comment (aaronlahey) [inline Server.Core/Server.cs]: Not sure `BuildRequestFromWholeMessage` is descriptive enough. Is the "whole message" included in a string, a byte[], a stream, etc...?

**Review by aaronlahey — CHANGES_REQUESTED** [review]

File: Server.Core/RequestBuilder.cs (hunk 3 above)
Comment (k-owens) [reply Server.Core/RequestBuilder.cs]: This method is only used when the whole message is given as a byte[].  In all of those cases startIndex is always different.  Based on your question I think I may have a naming issue.

File: Server.Core/RequestBuilder.cs (hunk 4 above)
Comment (k-owens) [reply Server.Core/RequestBuilder.cs]: I'm a little confused about your second point.  What do you mean by the parsing needn't be public to the world? Everything that uses this API doesn't know about the way it creates the requests and all of the parsing methods are private.

File: Server.Core/Response.cs (hunk 5 above)
Comment (k-owens) [reply Server.Core/Response.cs]: Hmm... you are right, but I am not sure which is wrong at the moment.  I'll have to think about that one.

File: Server.Core/Server.cs (hunk 6 above)
Comment (k-owens) [reply Server.Core/Server.cs]: Yes...

**Pull request merged**